### ISSUE TITLE
Question about loss in PP

### ISSUE THREAD

**Issue by zqwenn** [pytorch/torchtitan#931]
When I try to run the llama3_8b model, the loss is always -1. Are there any known issues related to this?

The train log is as below:
```
bash run_train.sh 
+ NGPU=8
+ LOG_RANK=0
+ CONFIG_FILE=./torchtitan/models/llama/train_configs/llama3_8b.toml
+ overrides=
+ '[' 0 -ne 0 ']'
+ PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
+ torchrun --nproc_per_node=8 --rdzv_backend c10d --rdzv_endpoint=localhost:0 --local-ranks-filter 0 --role rank --tee 3 torchtitan/train.py --job.config_file ./torchtitan/models/llama/train_configs/llama3_8b.toml
W0305 07:36:08.276310 1957 site-packages/torch/distributed/run.py:792] 
W0305 07:36:08.276310 1957 site-packages/torch/distributed/run.py:792] *****************************************
W0305 07:36:08.276310 1957 site-packages/torch/distributed/run.py:792] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0305 07:36:08.276310 1957 site-packages/torch/distributed/run.py:792] *****************************************
[rank0]:2025-03-05 07:36:12.140910: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
[rank0]:2025-03-05 07:36:12.155010: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
[rank0]:WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
[rank0]:E0000 00:00:1741160172.172059    2024 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
[rank0]:E0000 00:00:1741160172.177188    2024 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
[rank0]:2025-03-05 07:36:12.194883: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
[rank0]:To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
[rank0]:[titan] 2025-03-05 07:36:13,639 - root - INFO - Starting job: Llama 3 8B training
[rank0]:[titan] 2025-03-05 07:36:13,639 - root - INFO - [GC] Initial GC collection. 0.00 seconds.
[rank0]:[titan] 2025-03-05 07:36:13,847 - root - WARNING - ENV[TORCH_NCCL_ASYNC_ERROR_HANDLING] = 1 will be overridden to 3 based on job config
[rank0]:[titan] 2025-03-05 07:36:13,849 - root - INFO - CUDA capacity: NVIDIA A100-SXM4-80GB with 79.14GiB memory
[rank0]:[titan] 2025-03-05 07:36:14,155 - root - INFO - Peak FLOPS used for computing MFU: 3.120e+14
[rank0]:[titan] 2025-03-05 07:36:14,155 - root - INFO - Building 2-D device mesh with ['pp', 'dp_shard'], [4, 2]
[rank0]:[titan] 2025-03-05 07:36:21,094 - root - INFO - TikTokenizer built: #words 128256, BOS ID 128000, EOS ID 128001
[rank0]:[titan] 2025-03-05 07:36:21,094 - root - INFO - Preparing c4 dataset from /data/home/z30015011/titan_llama_datasets/c4/
[rank0]:[titan] 2025-03-05 07:36:21,646 - root - INFO - Building llama3 8B with TransformerModelArgs(_enforced='This field is used to enforce all fields have defaults.', dim=4096, n_layers=32, n_heads=32, n_kv_heads=8, vocab_size=128256, multiple_of=1024, ffn_dim_multiplier=1.3, norm_eps=1e-05, rope_theta=500000, max_seq_len=8192, depth_init=True, norm_type='rmsnorm')
[rank0]:[titan] 2025-03-05 07:36:22,002 - root - INFO - Model llama3 8B size: 8,030,261,248 total parameters
[rank0]:[titan] 2025-03-05 07:36:22,002 - root - INFO - No 'pipeline_parallel_split_points' provided so the generated splits are: ['layers.8', 'layers.16', 'layers.24'] This may be sub-optimal as the number of layers per stage may be unbalanced.
[rank0]:[titan] 2025-03-05 07:36:22,030 - root - INFO - PP rank 0 is building stage_idx 0 with start_layer None, stop_layer layers.8
[rank0]:[titan] 2025-03-05 07:36:22,030 - root - INFO - Using pipeline schedule 1F1B with 4 microbatches and 4 stages.
[rank0]:[titan] 2025-03-05 07:36:22,031 - root - INFO - Applied selective activation checkpointing to the model
[rank0]:[titan] 2025-03-05 07:36:22,076 - root - INFO - Applied FSDP to the model
[rank0]:[titan] 2025-03-05 07:36:22,393 - root - INFO - CUDA memory usage for model: 4.26GiB(5.38%)
[rank0]:[titan] 2025-03-05 07:36:22,395 - root - INFO - Training starts at step 1, with local batch size 4, global batch size 8, sequence length 8192, total steps 20 (warmup 200)
[rank0]:[titan] 2025-03-05 07:36:22,395 - root - INFO - Profiling active. Traces will be saved at ./outputs/profile_trace
[rank0]:[rank0]:[W305 07:36:23.543253190 ProcessGroupNCCL.cpp:3436] Warning: TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (function operator())
[rank0]:[rank0]:[W305 07:36:33.472562901 ProcessGroupNCCL.cpp:3436] Warning: TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (function operator())
[rank0]:[titan] 2025-03-05 07:36:38,701 - root - INFO - step:  1  loss: -1.0000  memory: 37.67GiB(47.60%)  tps: 502  tflops: 29.10  mfu: 9.33%
[rank0]:[titan] 2025-03-05 07:36:38,701 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:[titan] 2025-03-05 07:37:22,672 - root - INFO - step: 10  loss: -1.0000  memory: 47.07GiB(59.47%)  tps: 1,677  tflops: 97.11  mfu: 31.12%
[rank0]:[titan] 2025-03-05 07:38:11,636 - root - INFO - step: 20  loss: -1.0000  memory: 47.07GiB(59.47%)  tps: 1,673  tflops: 96.90  mfu: 31.06%
[rank0]:[titan] 2025-03-05 07:38:11,636 - root - INFO - Sleeping 2 seconds for other ranks to complete
[rank0]:[titan] 2025-03-05 07:38:13,637 - root - INFO - Training completed

```

**Comment by lessw2020** [comment]
HI @zqwenn  - this is b/c for most PP schedules, the loss is only output on the last rank (PP_len -1) vs for FSDP and similar, it's shown on rank 0.
If you adjust the logging to include your last rank (in the run_llama_train, it has log_rank=0 so add ,7 if you are on a single node with 8 gpus), then you'lll get your loss showing.

**Comment by lessw2020** [comment]
Note that while I say most, for the newest schedules which are V block based, then the loss shows again on 0.  
Also the loss is reported in the metrics (ala tensor board or W&B) correclty b/c there the loss output is adjusted...but then breaks for V schedule.
Anyway, we ned to get this more robustly fixed but for the console aspect, the fix atm is to include the last rank in your output!
Hope this helps!
